audformat/core/utils.py

```diff
@@ -1074,3 +1074,52 @@ def union(
     index = index.drop_duplicates()
 
     return index
+
+
+def iter_by_file(
+        obj: typing.Union[
+            pd.Index,
+            pd.Series,
+            pd.DataFrame,
+        ],
+) -> (str, typing.Iterator[typing.Union[pd.Index, pd.Series, pd.DataFrame]]):
+    r"""Iterate over index by file.
+
+    In each iteration return sub-index for one file.
+
+    Args:
+        obj: Can be a DataFrame or Series that index is conform to
+            :ref:`table specifications <data-tables:Tables>`, or just
+            the index object itself.
+
+    Returns:
+        If obj is an index, the return value will be an iterator in form of
+        (file, sub_index) if the index is segmentwise otherwise
+        (file, filewise_index(file)). If obj is a Series or a Dataframe, the
+        return value will be an iterator in form of (file, sub_obj).
+
+    Examples
+        >>> index = segmented_index(['f1', 'f1', 'f2'], [0, 1, 0], [2, 3, 1])
```

Comment:
I would suggest start with the filewise index and show the segmented index afterwards. And maybe also add an example for a non-index input, e.g.:

```
        >>> index = filewise_index(['f1', 'f1', 'f2'])
        >>> next(iter_by_file(index))
        ('f1', Index(['f1'], dtype='object', name='file'))
        >>> index = segmented_index(['f1', 'f1', 'f2'], [0, 1, 0], [2, 3, 1])
        >>> next(iter_by_file(index))
        ('f1', MultiIndex([('f1', '0 days 00:00:00', '0 days 00:00:02'),
            ('f1', '0 days 00:00:01', '0 days 00:00:03')],
           names=['file', 'start', 'end']))
        >>> obj = pd.Series(['a', 'b', 'b'], index)
        >>> next(iter_by_file(obj))
        ('f1', file  start            end
        f1    0 days 00:00:00  0 days 00:00:02    a
              0 days 00:00:01  0 days 00:00:03    b
        dtype: object)
```